### DISCUSSION TITLE
'dict' object has no attribute 'pk' when using HyperlinkedIdentityField with optimized queryset

### DISCUSSION THREAD

**Discussion by Guilouf** [encode/django-rest-framework#9430] [Potential Issue]
`HyperlinkedIdentityField` may not work with optimized querysets using `.values()`.

Let our model be a Car: 
```python
class Car(models.Model):
    brand = models.Charfield(max_length=50)
```
Our Serializer:
```python
class CarSerializer(ModelSerializer):
    api_url = HyperlinkedIdentityField(view_name='api:car-detail')

    class Meta:
        model = Car
        fields = ['api_url']
```
and viewset:
```python
class CarViewSet(viewsets.ReadOnlyModelViewSet):

    def get_queryset(self):
        return Car.objects.all().values('pk')
```
then i obtain this error `'dict' object has no attribute 'pk'` because the `get_url`https://github.com/encode/django-rest-framework/blob/fbdab09c776d5ceef041793a7acd1c9e91695e5d/rest_framework/relations.py#L321 method of `HyperlinkedIdentityField` is expecting a model instance, not a dict.
Tested with DRF 3.15.1

Seems to be an issue to me, but i followed the guidelines and asked it here first. What do you think .
If i get your approval i will go on and create an issue, i can even work on a PR.

**Comment by browniebroke** [comment]
Have you tried changing the queryset optimisation to [`.only()`](https://docs.djangoproject.com/en/5.0/ref/models/querysets/#only)? 

```python
class CarViewSet(viewsets.ReadOnlyModelViewSet):

    def get_queryset(self):
        return Car.objects.all().only('pk')
```

I think it would have the same effect as `.values()` from the SQL perspective, in both cases, your SQL looks linke this:

```sql
SELECT "vehicules_car"."id" FROM "vehicules_car"
```

The resulting queryset will contain model instances, instead of dicts.

**Reply by Guilouf** [reply]
I was writting my comment bellow in the same time... `only()` is much slower in my case.  This is my real queryset:
```python
return Accident.objects.select_related(
                'commune_fk'
            ).filter(
                commune_fk__isnull=False,
                # departement=85,  # fixme remove
            ).annotate(
                commune_center=AsGeoJSON('commune_fk__centroid')
            ).values(
                'pk',
                'commune_center',
            )
```
With only, i had to include `commune_fk` to get the annotation, and it's much slower. Anyway, everithing was fine using `values()`, just not `HyperlinkedIdentityField`, so it should be consistent ?

**Comment by Guilouf** [comment]
In my situation (more complex than this example), using `.only()` instead of `.values()` increase response time by *10, so it could not be a solution either. The only alternative i found is using `SerializerMethodField`, like this:

```python
    def get_api_url(self, obj):
        request = self.context.get('request')
        abs_url = reverse('api:cars-detail', kwargs={"pk": obj['pk']})
        return request.build_absolute_uri(abs_url)
```

**Reply by Guilouf** [reply]
Yet again, it's almost 4 time slower than when i just return the primary key, but that's another story.

**Reply by browniebroke** [reply]
Ok I see. Using `SerializerMethodField` is probably your best bet here. I don't think that `HyperlinkedIdentityField` is intended to work with `values` and at this point of DRF, this is unlikely to implement such change (see https://github.com/encode/django-rest-framework/discussions/9130 and https://github.com/encode/django-rest-framework/discussions/9270). 

When a use case like your is too specific for this field to handle, there is a reasonable escape hatch. If you want to go one step further, you could easily subclass `HyperlinkedIdentityField` in your own project to make it work with `dict`/`values`. Doing it in DRF is on order of magnitude harder due to the number of projects that depend on the existing behaviour.

**Reply by Guilouf** [reply]
Ok thanks, i thought rewritting `get_url` by something like
```python
    def get_url(self, obj, view_name, request, format):
        # Unsaved objects will not yet have a valid URL.
        if hasattr(obj, 'pk') and obj.pk in (None, ''):
            return None

        if isinstance(obj, dict):  # if .values() is used
            lookup_value = obj[self.lookup_field]
        else:
            lookup_value = getattr(obj, self.lookup_field)

        kwargs = {self.lookup_url_kwarg: lookup_value}
        return self.reverse(view_name, kwargs=kwargs, request=request, format=format)
```
won't introduce any breaking change and could make HyperlinkedIdentityField work with values.

**Comment by Guilouf** [comment]
I someone wondered how to include the object url without slowing down the api here is how i did it:
```python
    def get_queryset(self):
        """Queryset is optimized for list or retrieve views"""
            magic_key = 'magic_key'
            abs_url = reverse('api:accidents-detail', kwargs={"pk": 'magic_key'})
            full_url = self.request.build_absolute_uri(abs_url)

            return Car.objects.annotate(
                api_url=Replace(Value(full_url), Value(magic_key), Cast(F('pk'), output_field=CharField()), output_field=CharField())
            ).values(
                'api_url',
            )
```
I build the url wich is a rather slow process only once, then proceed in SQL to replace the part corresponding to the actual objet id.
